disnake/ext/commands/params.py

```diff
@@ -177,6 +175,88 @@ def register(cls, function: CallableT, annotation: Any) -> CallableT:
         return function
 
 
+class RangeMeta(type):
+    """Custom Generic implementation for Range"""
+
+    @overload
+    def __getitem__(self, args: Tuple[Union[int, ellipsis], Union[int, ellipsis]]) -> Type[int]:
+        ...
+
+    @overload
+    def __getitem__(
+        self, args: Tuple[Union[float, ellipsis], Union[float, ellipsis]]
+    ) -> Type[float]:
+        ...
+
+    def __getitem__(self, args: Tuple[Any, ...]) -> Any:
+        a, b = [None if x is Ellipsis else x for x in args]
+        return Range.create(min_value=a, max_value=b)
+
+
+class Range(type, metaclass=RangeMeta):
+    """Type depicting a limited range of allowed values"""
+
+    min_value: Optional[float]
+    max_value: Optional[float]
+
+    @overload
+    @classmethod
+    def create(
+        cls,
+        min_value: int = None,
+        max_value: int = None,
+        *,
+        le: int = None,
+        lt: int = None,
+        ge: int = None,
+        gt: int = None,
+    ) -> Type[int]:
+        ...
+
+    @overload
+    @classmethod
+    def create(
+        cls,
+        min_value: float = None,
+        max_value: float = None,
+        *,
+        le: float = None,
+        lt: float = None,
+        ge: float = None,
+        gt: float = None,
+    ) -> Type[float]:
+        ...
+
+    @classmethod
+    def create(
+        cls,
+        min_value: float = None,
+        max_value: float = None,
+        *,
+        le: float = None,
+        lt: float = None,
+        ge: float = None,
+        gt: float = None,
+    ) -> Any:
+        """Construct a new range with any possible constraints"""
+        self = cls(cls.__name__, (), {})
+        self.min_value = min_value if min_value is not None else _xt_to_xe(le, lt, -1)
+        self.max_value = max_value if max_value is not None else _xt_to_xe(ge, gt, 1)
```

Comment:
Why not something like `cls(min_value, max_value)` ?

Reply:
We inherit from `type` because only types may be used in special type annotation subscripts. (for example `Optional[Range[1, 10]]`)

Overwriting the `__new__` of a `type` subclass may cause unneeded issues.